### ISSUE TITLE
Workers stop training if number of clients partificpating in every round is changed from the default value

### ISSUE THREAD

**Issue by scngupta-dsp** [ssg-research/WAFFLE#6]
ISSUE: 
======
Workers stop training (avg loss: NaN) if number of clients partificpating in every round (Csub) is changed from the default value

STEPS TO REPRODUCE THE ISSUE
============================

1. WAFFLE/src/experiment.py, 
    class Experiment(object):
	def __init__(self, environment):
			
	- Change self.real_client_num = 20
			
2. Run command: 
    python main.py --config_file configurations/baseline_experiments/mnist/1.ini --experiment training
		   
3. Results: 

		[DEFAULT]
		train_batch_size: 64
		test_batch_size: 64
		dataset_name: MNIST
		epochs: 250
		local_epoch: 1
		client_num: 100
		input_size: 28
		number_of_classes: 10
		model_save_path: data/models/
		scores_save_path: train_record/
		checkpoint_save_path: checkpoint/
		use_cuda: 1
		iid: 1

		[FEDERATED]
		model_name: federated_mnist_l5_100c_1localround.pt
		train_dataset_save_path: data/datasets/MNIST/
		retrain: 0
		model_architecture: MNIST_L5
		resume: 0

		[WATERMARK]
		model_name: federated_mnist_l5_100c_1localround.pt
		normalize_with_imagenet_vals: 0
		decay: 0.00005
		pretrain: 0
		pretrain_epochs: 25
		w_retrain: 0
		watermark_model_save_path: data/watermark_model/
		watermark_size: 100
		learning_rate: 0.1
		force_greyscale: 1
		watermark_dataset_save_path: ./data/datasets/MWAFFLE/
		watermark_set: PATTERN
		model_architecture: MNIST_L5
		resume: 0
		>>> WARN: train_record/ does not exist. Creating...
		>>> INFO: train_record/ Created.
		>>> INFO: Training (MNIST) samples: 60000
		Test samples: 10000
		Saved in: data/datasets/MNIST/
		>>> INFO: Watermark (PATTERN) samples: 100
		Saved in: ./data/datasets/MWAFFLE/
		Worker:32 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:18.33
		Worker:98 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:15.83
		Worker:41 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:19.17
		Worker:94 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:23.00
		Worker:0 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:21.33
		Worker:61 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:20.67
		Worker:26 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:25.67
		Worker:19 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:21.17
		Worker:25 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:23.17
		Worker:75 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:17.67
		Worker:17 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:22.17
		Worker:40 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:18.67
		Worker:48 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:18.17
		Worker:83 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:18.83
		Worker:24 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:15.83
		Worker:42 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:19.83
		Worker:77 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:21.50
		Worker:18 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:21.83
		Worker:60 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:19.50
		Worker:28 Train Epoch: 1, worker data size 600, worker avg loss: 0.046, worker train acc:22.00
		>>> INFO: Accuracy of the network on the 100 watermark images (10/100): 10.0
		>>> INFO: Test accuracy: (2989/10000) 29.89%
		>>> INFO: Accuracy of the network on the 10000 test images (per-class):
		Accuracy of     0 : 46 %
		Accuracy of     1 : 99 %
		Accuracy of     2 : 40 %
		Accuracy of     3 : 93 %
		Accuracy of     4 :  0 %
		Accuracy of     5 :  0 %
		Accuracy of     6 :  0 %
		Accuracy of     7 :  0 %
		Accuracy of     8 :  4 %
		Accuracy of     9 :  1 %
		Testing at 1th epoch, train loss:0.009160764729976654
		>>> INFO: Test accuracy: (2989/10000) 29.89%
		>>> INFO: Accuracy of the network on the 10000 test images (per-class):
		Accuracy of     0 : 46 %
		Accuracy of     1 : 99 %
		Accuracy of     2 : 40 %
		Accuracy of     3 : 93 %
		Accuracy of     4 :  0 %
		Accuracy of     5 :  0 %
		Accuracy of     6 :  0 %
		Accuracy of     7 :  0 %
		Accuracy of     8 :  4 %
		Accuracy of     9 :  1 %
		Worker:43 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:11.17
		Worker:15 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:11.83
		Worker:2 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:13.17
		Worker:81 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:10.50
		Worker:93 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:12.00
		Worker:74 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:11.17
		Worker:13 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:11.67
		Worker:68 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:12.83
		Worker:12 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:12.17
		Worker:86 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:12.17
		Worker:95 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:11.33
		Worker:73 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:10.00
		Worker:84 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:11.50
		Worker:11 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:12.50
		Worker:18 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:13.33
		Worker:70 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:12.33
		Worker:78 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:11.50
		Worker:60 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:11.50
		Worker:44 Train Epoch: 2, worker data size 600, worker avg loss: inf, worker train acc:15.83
		Worker:50 Train Epoch: 2, worker data size 600, worker avg loss: nan, worker train acc:10.83
		>>> INFO: Accuracy of the network on the 100 watermark images (10/100): 10.0
		>>> INFO: Test accuracy: (980/10000) 9.80%
		>>> INFO: Accuracy of the network on the 10000 test images (per-class):
		Accuracy of     0 : 99 %
		Accuracy of     1 :  0 %
		Accuracy of     2 :  0 %
		Accuracy of     3 :  0 %
		Accuracy of     4 :  0 %
		Accuracy of     5 :  0 %
		Accuracy of     6 :  0 %
		Accuracy of     7 :  0 %
		Accuracy of     8 :  0 %
		Accuracy of     9 :  0 %
		Testing at 2th epoch, train loss:nan

**Comment by scngupta-dsp** [comment]
I suspect there is a problem while updating weights in calculate_weight function

		def calculate_weight(updated_w, client_subset, num_clients, model):
			with torch.set_grad_enabled(False):
				for i, (name, param) in enumerate(model.named_parameters(), 0):
					updated_w[i] += (client_subset * param.get())/num_clients
				

1. In the default case 
    - client_subset = 10 & num_clients   = 100
	This results in every client weight being weighed by 0.1 (10/100)  in updated_w
	
2. In modified case: 
    Every client weight is being weighed by 0.2 and seems weight aggregation is going wrong 
	
The issue can be solved by modifying the logic as follows: 

		def calculate_weight(updated_w, client_subset, num_clients, model):
			with torch.set_grad_enabled(False):
				for i, (name, param) in enumerate(model.named_parameters(), 0):
					updated_w[i] += param.get())/client_subset

**Comment by scngupta-dsp** [comment]
I suspect there is a problem while updating weights in calculate_weight function

		def calculate_weight(updated_w, client_subset, num_clients, model):
			with torch.set_grad_enabled(False):
				for i, (name, param) in enumerate(model.named_parameters(), 0):
					updated_w[i] += (client_subset * param.get())/num_clients
				

1. In the default case 
    - client_subset = 10 & num_clients   = 100
	This results in every client weight being weighed by 0.1 (10/100)  in updated_w
	
2. In modified case: 
    Every client weight is being weighed by 0.2 and seems weight aggregation is going wrong 
	
The issue can be solved by modifying the logic as follows: 

		def calculate_weight(updated_w, client_subset, num_clients, model):
			with torch.set_grad_enabled(False):
				for i, (name, param) in enumerate(model.named_parameters(), 0):
					updated_w[i] += param.get())/client_subset

**Comment by bussfromspace** [comment]
Hi,

Thanks for catching the bug. You are right, the logic should be like above to make federated averaging correct for different client subsets. This correction do not change results in paper, but solves NaN issue.
